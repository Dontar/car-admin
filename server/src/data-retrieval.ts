import { getDB } from "./connection";
import rawSql, { in as $in, select } from 'sql-bricks-sqlite';
import { GetOneParams, GetOneResult, GetParams, GetResult, ICars, ICompany, IPerson } from "./share/models";

export async function getCars(params: GetOneParams): Promise<GetOneResult<ICars>>;
export async function getCars(params: Partial<GetParams>): Promise<GetResult<ICars>>;
export async function getCars(params: Partial<GetParams> | GetOneParams): Promise<GetResult<ICars> | GetOneResult<ICars>> {
    const db = await getDB();

    const sql = select(
        'v.id as id',
        'upper(v.rama) as rama',
        'v.dkn as dkm',
        'vm.name as mark_name',
        'vmd.name as model_name',
        'v.produce_year as produce_year',
        'max(vc.client_id) as company_id',
        'max(vc.client_id) as person_id'
    ).from('vehicles v')
        .join('vehicle_mark vm', { 'vm.ID': 'v.MARK_ID' })
        .join('vehicle_model vmd', { 'vmd.ID': 'v.MODEL_ID' })
        .leftJoin('vc', { 'vc.VEHICLE_ID': 'v.ID' })
        .groupBy('v.ID');

    let total: number | undefined = 1376287;
    if (isOne(params)) {
        sql.where({ 'v.id': params.id });
    } else {
        if (params.ids) {
            sql.where($in('v.id', ...params.ids));
        } else {
            /* total = */ await prepareParams(params, sql, db);
        }
    }

    const { text, values } = sql.toParams();
    if (isOne(params)) {
        const row = await db.get<ICars>(text, values);
        return {
            data: row!
        }
    } else {
        const rows = await db.all<ICars[]>(text, values);
        return {
            data: rows,
            total
        }
    }
}

export async function getPersons(params: GetOneParams): Promise<GetOneResult<IPerson>>;
export async function getPersons(params: Partial<GetParams>): Promise<GetResult<IPerson>>;
export async function getPersons(params: Partial<GetParams> | GetOneParams): Promise<GetResult<IPerson> | GetOneResult<IPerson>> {
    const db = await getDB();

    const sql = select(
        'c.ID as id',
        'c.CLIENT_NAME as client_name',
        'c.EGN as egn',
        'c.REPRESENTATIVE as representative',
        'c.PHONE as phone',
        'c.MOBILE as mobile',
        'c.REGION_NAME as region_name',
        'c.MINUCIPALITY_NAME as municipality_name',
        'c.CITY as city',
        'c.POSTCODE as postcode',
        'c.STREET_NAME as street_name',
        'c.STREET_NO as street_no',
        'c.BLOK as blok',
        'c.VHOD as vhod',
        'c.APARTMENT as apartment',
        'c.FLOOR as floor',
        'group_concat(vc.VEHICLE_ID) as car_ids'
    ).from('clients c')
        .leftJoin('vc', { 'vc.CLIENT_ID': 'c.ID' })
        .where('c.CLIENT_TYPE', rawSql(1))
        .groupBy('c.ID');

    let total: number | undefined = 1030889;
    if (isOne(params)) {
        sql.where({ 'c.id': params.id });
    } else {
        if (params.ids) {
            sql.where($in('c.id', ...params.ids));
        } else {
            /* total = */ await prepareParams(params, sql, db);
        }
    }

    const { text, values } = sql.toParams();
    if (isOne(params)) {
        const row = await db.get<IPerson>(text, values);
        return {
            data: stringIdsToArray(row!)
        };
    } else {
        const rows = await db.all<IPerson[]>(text, values);
        return {
            data: stringIdsToArray(rows),
            total
        };
    }
}

export async function getCompanies(params: GetOneParams): Promise<GetOneResult<ICompany>>;
export async function getCompanies(params: Partial<GetParams>): Promise<GetResult<ICompany>>;
export async function getCompanies(params: Partial<GetParams> | GetOneParams): Promise<GetResult<ICompany> | GetOneResult<ICompany>> {
    const db = await getDB();

    const sql = select(
        'c.ID as id',
        'c.CLIENT_NAME as client_name',
        'c.EGN as bulstat',
        'c.REPRESENTATIVE as representative',
        'c.PHONE as phone',
        'c.MOBILE as mobile',
        'c.REGION_NAME as region_name',
        'c.MINUCIPALITY_NAME as municipality_name',
        'c.CITY as city',
        'c.POSTCODE as postcode',
        'c.STREET_NAME as street_name',
        'c.STREET_NO as street_no',
        'c.BLOK as blok',
        'c.VHOD as vhod',
        'c.APARTMENT as apartment',
        'c.FLOOR as floor',
        'group_concat(vc.VEHICLE_ID) as car_ids'
    ).from('clients c')
        .leftJoin('vc', { 'vc.CLIENT_ID': 'c.ID' })
        .where('c.CLIENT_TYPE', rawSql(2))
        .groupBy('c.ID');

    let total: number | undefined = 78689;
    if (isOne(params)) {
        sql.where({ 'c.id': params.id });
    } else {
        if (params.ids) {
            sql.where($in('c.id', ...params.ids));
        } else {
            /* total = */ await prepareParams(params, sql, db);
        }
    }

    const { text, values } = sql.toParams();
    if (isOne(params)) {
        const row = await db.get<ICompany>(text, values);
        return {
            data: stringIdsToArray(row!)
        }
    } else {
        const rows = await db.all<ICompany[]>(text, values);
        return {
            data: stringIdsToArray(rows),
            total
        }
    }
}

function isOne(params: any): params is GetOneParams {
    return params.id !== null && params.id !== undefined;
}

async function prepareParams(params: Partial<GetParams>, sql: rawSql.SelectStatement, db: any) {
    let total: number | undefined = undefined;
    // const { text, values } = select('COUNT(*) as total').from(sql).toParams();
    // total = (await db.get(text, values))!.total;

    if (params.target) {
        sql.where(params.target, rawSql(params.id));
    }

    if (params.sort) {
        const { field, order } = params.sort;
        sql.orderBy(field + ' ' + order);
    }

    if (params.pagination) {
        const { page, perPage } = params.pagination;
        sql.limit(perPage);
        sql.offset((page - 1) * perPage);
    }

    return total;
}

function stringIdsToArray<T>(obj: T): T;
function stringIdsToArray<T>(obj: T | T[]): T | T[] {

    function transform<T>(obj: T): T {
        Object.keys(obj).forEach(field => {
            if (field.substr(field.length - 4) === '_ids') {
                if (typeof (<any>obj)[field] === 'string') {
                    const val: string = (<any>obj)[field];
                    (<any>obj)[field] = val.split(',').map(i => Number(i));
                }
            }
        });
        return obj;
    }

    if (Array.isArray(obj)) {
        return obj.map(obj => transform(obj));
    } else {
        return transform(obj);
    }
}