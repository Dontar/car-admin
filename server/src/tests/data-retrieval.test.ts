import { getCars, getCompanies, getPersons } from "../data-retrieval";
import { config } from 'dotenv';
config();

jest.setTimeout(5000 * 10);

describe('Test car data retrieval', () => {

    it('should get a car from db', async () => {
        const data = await getCars({
            id: 4601
        });
        expect(data).toBeDefined();
        expect(data).toMatchSnapshot();
    });

    it('should get 2 cars from db', async () => {
        const data = await getCars({
            ids: [4601, 4602]
        });
        expect(data.data?.length).toBe(2);
        expect(data).toMatchSnapshot();
    });

    it('should get cars from db with pagination', async () => {
        const data = await getCars({
            pagination: { page: 1, perPage: 2 }
        });
        expect(data.data?.length).toBe(2);
        expect(data).toMatchSnapshot();
    });

    it('should get cars from db with sorting', async () => {
        const data = await getCars({
            pagination: { page: 1, perPage: 2 },
            sort: { field: 'id', order: 'ASC' }
        });
        expect(data.data?.length).toBe(2);
        expect(data).toMatchSnapshot();
    });

});

describe('Test companies data retrieval', () => {

    it('should get a company from db', async () => {
        const data = await getCompanies({
            id: 9351
        });
        expect(data).toBeDefined();
        expect(data).toMatchSnapshot();
    });

    it('should get 2 companies from db', async () => {
        const data = await getCompanies({
            ids: [9351, 9851]
        });
        expect(data.data?.length).toBe(2);
        expect(data).toMatchSnapshot();
    });

    it('should get companies from db with pagination', async () => {
        const data = await getCompanies({
            pagination: { page: 1, perPage: 2 }
        });
        expect(data.data?.length).toBe(2);
        expect(data).toMatchSnapshot();
    });

    it('should get companies from db with sorting', async () => {
        const data = await getCompanies({
            pagination: { page: 1, perPage: 2 },
            sort: { field: 'id', order: 'ASC' }
        });
        expect(data.data?.length).toBe(2);
        expect(data).toMatchSnapshot();
    });
});

describe('Test persons data retrieval', () => {

    it('should get a person from db', async () => {
        const data = await getPersons({
            id: 9219
        });
        expect(data).toBeDefined();
        expect(data).toMatchSnapshot();
    });

    it('should get 2 persons from db', async () => {
        const data = await getPersons({
            ids: [9219, 9451]
        });
        expect(data.data?.length).toBe(2);
        expect(data).toMatchSnapshot();
    });

    it('should get persons from db with pagination', async () => {
        const data = await getPersons({
            pagination: { page: 1, perPage: 2 }
        });
        expect(data.data?.length).toBe(2);
        expect(data).toMatchSnapshot();
    });

    it('should get persons from db with sorting', async () => {
        const data = await getPersons({
            pagination: { page: 1, perPage: 2 },
            sort: { field: 'id', order: 'ASC' }
        });
        expect(data.data?.length).toBe(2);
        expect(data).toMatchSnapshot();
    });
});