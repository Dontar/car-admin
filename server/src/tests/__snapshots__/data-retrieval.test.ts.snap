// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test car data retrieval should get 2 cars from db 1`] = `
Object {
  "data": Array [
    Object {
      "client_id": null,
      "dkm": "CA4918KM",
      "id": 4601,
      "mark_name": "Volkswagen",
      "model_name": "Passat",
      "produce_year": 1992,
      "rama": "YGHJHKLLOLIHGHG",
    },
    Object {
      "client_id": null,
      "dkm": "CA7474KA",
      "id": 4602,
      "mark_name": "Audi",
      "model_name": "A8",
      "produce_year": 2008,
      "rama": "WVWEZZZ1HZRW206119",
    },
  ],
  "total": 1376287,
}
`;

exports[`Test car data retrieval should get a car from db 1`] = `
Object {
  "data": Object {
    "client_id": null,
    "dkm": "CA4918KM",
    "id": 4601,
    "mark_name": "Volkswagen",
    "model_name": "Passat",
    "produce_year": 1992,
    "rama": "YGHJHKLLOLIHGHG",
  },
}
`;

exports[`Test car data retrieval should get cars from db with pagination 1`] = `
Object {
  "data": Array [
    Object {
      "client_id": null,
      "dkm": "CA4918KM",
      "id": 4601,
      "mark_name": "Volkswagen",
      "model_name": "Passat",
      "produce_year": 1992,
      "rama": "YGHJHKLLOLIHGHG",
    },
    Object {
      "client_id": null,
      "dkm": "CA7474KA",
      "id": 4602,
      "mark_name": "Audi",
      "model_name": "A8",
      "produce_year": 2008,
      "rama": "WVWEZZZ1HZRW206119",
    },
  ],
  "total": 1376287,
}
`;

exports[`Test car data retrieval should get cars from db with sorting 1`] = `
Object {
  "data": Array [
    Object {
      "client_id": null,
      "dkm": "CA4918KM",
      "id": 4601,
      "mark_name": "Volkswagen",
      "model_name": "Passat",
      "produce_year": 1992,
      "rama": "YGHJHKLLOLIHGHG",
    },
    Object {
      "client_id": null,
      "dkm": "CA7474KA",
      "id": 4602,
      "mark_name": "Audi",
      "model_name": "A8",
      "produce_year": 2008,
      "rama": "WVWEZZZ1HZRW206119",
    },
  ],
  "total": 1376287,
}
`;

exports[`Test companies data retrieval should get 2 companies from db 1`] = `
Object {
  "data": Array [
    Object {
      "apartment": "0",
      "blok": "0",
      "bulstat": "175153616",
      "city": "ЛОЗЕН",
      "client_name": "Бонев Транс ЕООД  ",
      "floor": "0,00",
      "id": 9351,
      "mobile": "0",
      "municipality_name": "Столична",
      "phone": "0",
      "postcode": "1151,00",
      "region_name": "София (столица)",
      "representative": null,
      "street_name": "ЯНКО ПАНАЙОТОВ",
      "street_no": "28",
      "vehicle_ids": "1863595,1687957,1687957,1687957,6001,1687957,6001",
      "vhod": "0",
    },
    Object {
      "apartment": "27",
      "blok": "1",
      "bulstat": "130988423",
      "city": "СОФИЯ",
      "client_name": "\\\\ Мечкуеви Ко \\\\ ООД  ",
      "floor": "2,00",
      "id": 9851,
      "mobile": "0",
      "municipality_name": "Столична",
      "phone": "0",
      "postcode": "1000,00",
      "region_name": "София (столица)",
      "representative": null,
      "street_name": "Незабравка",
      "street_no": "1",
      "vehicle_ids": null,
      "vhod": "Б",
    },
  ],
  "total": 78689,
}
`;

exports[`Test companies data retrieval should get a company from db 1`] = `
Object {
  "data": Object {
    "apartment": "0",
    "blok": "0",
    "bulstat": "175153616",
    "city": "ЛОЗЕН",
    "client_name": "Бонев Транс ЕООД  ",
    "floor": "0,00",
    "id": 9351,
    "mobile": "0",
    "municipality_name": "Столична",
    "phone": "0",
    "postcode": "1151,00",
    "region_name": "София (столица)",
    "representative": null,
    "street_name": "ЯНКО ПАНАЙОТОВ",
    "street_no": "28",
    "vehicle_ids": "1863595,1687957,1687957,1687957,6001,1687957,6001",
    "vhod": "0",
  },
}
`;

exports[`Test companies data retrieval should get companies from db with pagination 1`] = `
Object {
  "data": Array [
    Object {
      "apartment": "0",
      "blok": "0",
      "bulstat": "175153616",
      "city": "ЛОЗЕН",
      "client_name": "Бонев Транс ЕООД  ",
      "floor": "0,00",
      "id": 9351,
      "mobile": "0",
      "municipality_name": "Столична",
      "phone": "0",
      "postcode": "1151,00",
      "region_name": "София (столица)",
      "representative": null,
      "street_name": "ЯНКО ПАНАЙОТОВ",
      "street_no": "28",
      "vehicle_ids": "1863595,1687957,1687957,1687957,6001,1687957,6001",
      "vhod": "0",
    },
    Object {
      "apartment": "27",
      "blok": "1",
      "bulstat": "130988423",
      "city": "СОФИЯ",
      "client_name": "\\\\ Мечкуеви Ко \\\\ ООД  ",
      "floor": "2,00",
      "id": 9851,
      "mobile": "0",
      "municipality_name": "Столична",
      "phone": "0",
      "postcode": "1000,00",
      "region_name": "София (столица)",
      "representative": null,
      "street_name": "Незабравка",
      "street_no": "1",
      "vehicle_ids": null,
      "vhod": "Б",
    },
  ],
  "total": 78689,
}
`;

exports[`Test companies data retrieval should get companies from db with sorting 1`] = `
Object {
  "data": Array [
    Object {
      "apartment": "0",
      "blok": "0",
      "bulstat": "175153616",
      "city": "ЛОЗЕН",
      "client_name": "Бонев Транс ЕООД  ",
      "floor": "0,00",
      "id": 9351,
      "mobile": "0",
      "municipality_name": "Столична",
      "phone": "0",
      "postcode": "1151,00",
      "region_name": "София (столица)",
      "representative": null,
      "street_name": "ЯНКО ПАНАЙОТОВ",
      "street_no": "28",
      "vehicle_ids": "1863595,1687957,1687957,1687957,6001,1687957,6001",
      "vhod": "0",
    },
    Object {
      "apartment": "27",
      "blok": "1",
      "bulstat": "130988423",
      "city": "СОФИЯ",
      "client_name": "\\\\ Мечкуеви Ко \\\\ ООД  ",
      "floor": "2,00",
      "id": 9851,
      "mobile": "0",
      "municipality_name": "Столична",
      "phone": "0",
      "postcode": "1000,00",
      "region_name": "София (столица)",
      "representative": null,
      "street_name": "Незабравка",
      "street_no": "1",
      "vehicle_ids": null,
      "vhod": "Б",
    },
  ],
  "total": 78689,
}
`;

exports[`Test persons data retrieval should get 2 persons from db 1`] = `
Object {
  "data": Array [
    Object {
      "apartment": "2",
      "blok": "18",
      "bulstat": "6310305565",
      "city": "СИЛИСТРА",
      "client_name": "First Middle Last",
      "floor": "1,00",
      "id": 9219,
      "mobile": "0891043493",
      "municipality_name": "Силистра",
      "phone": "02830123",
      "postcode": "7500,00",
      "region_name": "СИЛИСТРА",
      "representative": null,
      "street_name": "ГЕНЕРАЛ  СКОБЕЛЕВ",
      "street_no": "",
      "vehicle_ids": null,
      "vhod": "А",
    },
    Object {
      "apartment": "7",
      "blok": "9",
      "bulstat": "5507181593",
      "city": "СОФИЯ",
      "client_name": "Николинка Стефанова Цолова",
      "floor": "4,00",
      "id": 9451,
      "mobile": "0897527002",
      "municipality_name": "Столична",
      "phone": "0888998899",
      "postcode": "1000,00",
      "region_name": "София (столица)",
      "representative": null,
      "street_name": "ЛЮБЛАНА",
      "street_no": "1",
      "vehicle_ids": null,
      "vhod": "a",
    },
  ],
  "total": 1030889,
}
`;

exports[`Test persons data retrieval should get a person from db 1`] = `
Object {
  "data": Object {
    "apartment": "2",
    "blok": "18",
    "bulstat": "6310305565",
    "city": "СИЛИСТРА",
    "client_name": "First Middle Last",
    "floor": "1,00",
    "id": 9219,
    "mobile": "0891043493",
    "municipality_name": "Силистра",
    "phone": "02830123",
    "postcode": "7500,00",
    "region_name": "СИЛИСТРА",
    "representative": null,
    "street_name": "ГЕНЕРАЛ  СКОБЕЛЕВ",
    "street_no": "",
    "vehicle_ids": null,
    "vhod": "А",
  },
}
`;

exports[`Test persons data retrieval should get persons from db with pagination 1`] = `
Object {
  "data": Array [
    Object {
      "apartment": "2",
      "blok": "18",
      "bulstat": "6310305565",
      "city": "СИЛИСТРА",
      "client_name": "First Middle Last",
      "floor": "1,00",
      "id": 9219,
      "mobile": "0891043493",
      "municipality_name": "Силистра",
      "phone": "02830123",
      "postcode": "7500,00",
      "region_name": "СИЛИСТРА",
      "representative": null,
      "street_name": "ГЕНЕРАЛ  СКОБЕЛЕВ",
      "street_no": "",
      "vehicle_ids": null,
      "vhod": "А",
    },
    Object {
      "apartment": "7",
      "blok": "9",
      "bulstat": "5507181593",
      "city": "СОФИЯ",
      "client_name": "Николинка Стефанова Цолова",
      "floor": "4,00",
      "id": 9451,
      "mobile": "0897527002",
      "municipality_name": "Столична",
      "phone": "0888998899",
      "postcode": "1000,00",
      "region_name": "София (столица)",
      "representative": null,
      "street_name": "ЛЮБЛАНА",
      "street_no": "1",
      "vehicle_ids": null,
      "vhod": "a",
    },
  ],
  "total": 1030889,
}
`;

exports[`Test persons data retrieval should get persons from db with sorting 1`] = `
Object {
  "data": Array [
    Object {
      "apartment": "2",
      "blok": "18",
      "bulstat": "6310305565",
      "city": "СИЛИСТРА",
      "client_name": "First Middle Last",
      "floor": "1,00",
      "id": 9219,
      "mobile": "0891043493",
      "municipality_name": "Силистра",
      "phone": "02830123",
      "postcode": "7500,00",
      "region_name": "СИЛИСТРА",
      "representative": null,
      "street_name": "ГЕНЕРАЛ  СКОБЕЛЕВ",
      "street_no": "",
      "vehicle_ids": null,
      "vhod": "А",
    },
    Object {
      "apartment": "7",
      "blok": "9",
      "bulstat": "5507181593",
      "city": "СОФИЯ",
      "client_name": "Николинка Стефанова Цолова",
      "floor": "4,00",
      "id": 9451,
      "mobile": "0897527002",
      "municipality_name": "Столична",
      "phone": "0888998899",
      "postcode": "1000,00",
      "region_name": "София (столица)",
      "representative": null,
      "street_name": "ЛЮБЛАНА",
      "street_no": "1",
      "vehicle_ids": null,
      "vhod": "a",
    },
  ],
  "total": 1030889,
}
`;
